### PR TITLE
[Rel-Eng]: Adds script to automatically generate GHCup metadata for Cabal releases

### PR DESCRIPTION
**Pull request by arjunkathuria**
Draft PR for feedback and suggestion for issue #9298 

### The problem
Currently, on a cabal release, for it to be released by GHCup, the metadata of the release required by GHCup is manually generated. The process seems quite arduous and is an overhead for a release manager.

### The solution
This PR intends to introduce a script that would help make cabal releases easier and more streamlined by automatically generating the required cabal release metadata in the correct format that is usable by GHCup as-is. It then essentially becomes a simpler affair for the release manager to introduce the new release in GHCup.

The [link](url) in the issue 9298 thread was used as a reference for this script.

### Current Progress
The basic core functionality of the script works, as in, if you give it a release number as it's first and only argument, it generates the metadata for that release which can be used drop-in by GHCup. An example of the output to follow in the first comment of the PR.

More sections and info TBD

### Todos
- [ ] figure out the correct directory to CD into for artifacts/downloads [(the line here)](https://github.com/arjunkathuria/cabal/blob/release-metadata-script/scripts/release/create-release-metadata-for-ghcup.sh#L13)
- [ ] verify which directory the script should reside in. It currently resides in the top level `scripts/release` directory.
- [ ] TBD based on reviews and feedbacks.

### CODE REVIEW

**Comment by arjunkathuria** [comment]
The usage and current output result after executing the script can be seen as below:- 

### usage 

```
$ create-release-metadata-for-ghcup.sh "3.10.2.0"  
```
or, without the quotes 

```
$ create-release-metadata-for-ghcup.sh 3.10.2.0
```

both should work.


### The output

The script generates the metadata for the release in the following format, which can be used by GHCup drop-in:-


```
    3.10.2.0:
      viTags:
        - Latest
      viChangeLog: https://github.com/haskell/cabal/blob/master/release-notes/cabal-install-3.10.2.0.md
      viPostInstall: *cabal-31020-post-install
      viArch:
        A_64:
          Linux_UnknownLinux:
            unknown_versioning: &cabal-31020-64
            dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-x86_64-linux-alpine3_12.tar.xz
            dlSubdir: cabal-install-3.10.2.0
            dlHash: b54e1cd235c47c62c03cdb9f6cf90e5fe8ae38c5e5befb9e21c8d1395f4bde05
          Linux_Alpine:
            unknown_versioning: &cabal-31020-64
          Linux_CentOS:
            unknown_versioning: &cabal-31020-64-centos7
            dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-x86_64-linux-centos7.tar.xz
            dlSubdir: cabal-install-3.10.2.0
            dlHash: cfcdab399380dec7fedda55898bff975ac30b5d5d579433cbf8773b17c15f410
          Linux_Debian:
            ' ( >= 9 && < 10)': &cabal-31020-64-debian
              dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-x86_64-linux-deb9.tar.xz
              dlSubdir: cabal-install-3.10.2.0
              dlHash: af5ce42114cf7720c37fee3238781df4c75bb74914c62e6a68833fb434e0cad7
            ' ( == 10 && < 11)':
              dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-x86_64-linux-deb10.tar.xz
              dlSubdir: cabal-install-3.10.2.0
              dlHash: bdeb27c008b09c3b86f8a2768018d62a1aee02565304d123fda87ed432549418
            ' ( >= 11)':
              dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-x86_64-linux-deb11.tar.xz
              dlSubdir: cabal-install-3.10.2.0
              dlHash: 9ca5625c89e8fcada02edced5048c3a3db0254e2bef1eb792d549d633222b108
            unknown_versioning: &cabal-31020-64-debian
          Linux_Fedora:
            '>= 33':
              dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-x86_64-linux-fedora33.tar.xz
              dlSubdir: cabal-install-3.10.2.0
              dlHash: 1e59dc1e1a1b33085a1789b8ddafb55026211454efe0b4e814c956ef86fe3ea5
            unknown_versioning: &cabal-31020-64-centos7
          Linux_Ubuntu:
            '< 20': &cabal-31020-64-ubuntu18
              dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-x86_64-linux-ubuntu18_04.tar.xz
              dlSubdir: cabal-install-3.10.2.0
              dlHash: 7b0bb22c263ae0b43459de1995bf465560d412c12d47af011b2ac27b5d30c7aa
            '>= 20': &cabal-31020-64-ubuntu20
              dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-x86_64-linux-ubuntu20_04.tar.xz
              dlSubdir: cabal-install-3.10.2.0
              dlHash: c2a8048caa3dbfe021d0212804f7f2faad4df1154f1ff52bd2f3c68c1d445fe1
            unknown_versioning: *cabal-31020-64-ubuntu18
          Linux_Mint:
            '< 20': *cabal-31020-64-ubuntu18
            '>= 20': *cabal-31020-64-ubuntu20
            unknown_versioning: *cabal-31020-64-ubuntu18
          Darwin:
            unknown_versioning:
              dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-x86_64-darwin.tar.xz
              dlSubdir: cabal-install-3.10.2.0
              dlHash: cd64f2a8f476d0f320945105303c982448ca1379ff54b8625b79fb982b551d90
          Windows:
            unknown_versioning:
              dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-x86_64-windows.zip
              dlSubdir: cabal-install-3.10.2.0
              dlHash: b09e335b2ebeafa1db5e1e5614e3e10fb37da230a36865d76646ab27b2f3f46b
          FreeBSD:
            unknown_versioning:
              dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-x86_64-freebsd.tar.xz
              dlSubdir: cabal-install-3.10.2.0
              dlHash: 6dcd9d38a9f2101a0a3c3b74cacb2e41b8f7226f181780c0f872f2f1206dee37
        A_32:
          Linux_UnknownLinux:
            unknown_versioning: &cabal-31020-32
              dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-i386-linux-alpine3_12.tar.xz
              dlSubdir: cabal-install-3.10.2.0
              dlHash: e219d2c6446c29e09644c4f064f4b87d486871ad6f6de16f0d2fcbd8626b5a5b
          Linux_Alpine:
            unknown_versioning: *cabal-31020-32
          Linux_Debian:
            '( >= 9 )':
              dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-i386-linux-deb9.tar.xz
              dlSubdir: cabal-install-3.10.2.0
              dlHash: 2b26d2cb67f1ba3561509fbccc30810ccc1c5032238fca00666e3dcd03e941a8
            unknown_versioning: *cabal-31020-32
        A_ARM64:
          Darwin:
            unknown_versioning:
              dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-aarch64-darwin.tar.xz
              dlSubdir: cabal-install-3.10.2.0
              dlHash: d2bd336d7397cf4b76f3bb0d80dea24ca0fa047903e39c8305b136e855269d7b
          Linux_Debian:
            '( >= 10 && < 11)': &cabal-31020-arm64
              dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-aarch64-linux-deb10.tar.xz
              dlSubdir: cabal-install-3.10.2.0
              dlHash: 004ed4a7ca890fadee23f58f9cb606c066236a43e16b34be2532b177b231b06d
            '( >= 11)':
              dlUri: https://downloads.haskell.org/~cabal/cabal-install-3.10.2.0/cabal-install-3.10.2.0-aarch64-linux-deb11.tar.xz
              dlSubdir: cabal-install-3.10.2.0
              dlHash: daa767a1b844fbc2bfa0cc14b7ba67f29543e72dd630f144c6db5a34c0d22eb1
            unknown_versioning: *cabal-31020-arm64
          Linux_UnknownLinux:
            unknown_versioning: *cabal-31020-arm64
```

The output for release "3.10.2.0" from running this script is almost identical when compared to [what is currently deployed at GHCup for this release ](https://github.com/haskell/ghcup-metadata/blob/develop/ghcup-0.0.7.yaml#L4431-L4514) .

The hashes match and the urls are identical (except for the ~cabal part). 
There are certain differences like an addition of `dlSubdir` part in most sections.

**Comment by arjunkathuria** [comment]
It just occurred to me that they (@Kleidukos , @hasufell ) might have notifications off for draft PRs.
un-drafting for now.

**Comment by Kleidukos** [comment]
@arjunkathuria indeed, drafts are not meant to be reviewed. :)

**Comment by Kleidukos** [comment]
@arjunkathuria can you run `shellcheck` on your script?

**Comment by arjunkathuria** [comment]
@Kleidukos just did, came out clean.

image

File: scripts/release/create-release-metadata-for-ghcup.sh
Comment (Kleidukos) [inline scripts/release/create-release-metadata-for-ghcup.sh:1]:
```suggestion
#!/usr/bin/env bash
```

**Review by Kleidukos — CHANGES_REQUESTED** [review]
Only one change and it should be good :+1:

File: scripts/release/create-release-metadata-for-ghcup.sh
Comment (arjunkathuria) [reply scripts/release/create-release-metadata-for-ghcup.sh:1]: Sure.

Also wanted to verify if we'd need to `cd` into a directory for artifacts/downloads here -> https://github.com/arjunkathuria/cabal/blob/release-metadata-script/scripts/release/create-release-metadata-for-ghcup.sh#L12-L13

Comment (arjunkathuria) [reply scripts/release/create-release-metadata-for-ghcup.sh]: Done.

Comment (hasufell) [reply scripts/release/create-release-metadata-for-ghcup.sh]: > ```suggestion
> #!/usr/bin/env bash
> ```

Why? This isn't more portable. 

https://stackoverflow.com/a/53116875

**Comment by hasufell** [comment]
This script only works well in conjunction with a script that downloads artifacts into a specific dir.

See https://github.com/haskell/haskell-language-server/blob/master/scripts/release/download-gh-artifacts.sh

**Comment by arjunkathuria** [comment]
> This script only works well in conjunction with a script that downloads artifacts into a specific dir.
> 
> See https://github.com/haskell/haskell-language-server/blob/master/scripts/release/download-gh-artifacts.sh

Yup, that's what i was wondering, if we'd need to explicitly download them or run this script on a dir that already contains them somehow. 

I'll need to add that one too right ?

**Comment by hasufell** [comment]
The other option is to run it in CI, but that's not worth it imo. One has to download the artifacts either way to upload them to downloads.haskell.org. I don't think haskell infra allows automatic uploads through machine keys (I hope they don't), so it's a manual step.

**Comment by arjunkathuria** [comment]
I see.
Since we don't seem to do github binary releases for cabal-install, i suppose i'd have to download them from upstream links like https://downloads.haskell.org/~cabal/cabal-install-3.10.1.0/  ?

**Comment by arjunkathuria** [comment]
Added the script to download the binary release files and verify the downloaded checksums on them. To be used in conjunction with the GHCup metadata generating script.

File: scripts/release/download-cabal-install-release-binaries.sh
Comment (hasufell) [inline scripts/release/download-cabal-install-release-binaries.sh:31]: This is the wrong way around and would mean we download twice: once manually and once after we uploaded (why?).

The artifacts are built on gitlab CI. The release manager will have to download them to their local machine in order to upload them to downloads.haskell.org. And they want to perform the yaml step after they downloaded from gitlab, not after they uploaded to the download server.

Comment (arjunkathuria) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: > I see. Since we don't seem to do github binary releases for cabal-install, i suppose i'd have to download them from upstream links like https://downloads.haskell.org/~cabal/cabal-install-3.10.1.0/ ?

could've mentioned here when i asked to confirm.

What's the correct url here?

Comment (hasufell) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: @chreekat might know

Comment (arjunkathuria) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: 🕯️ * summons @chreekat * 🕯️

File: scripts/release/create-release-metadata-for-ghcup.sh
Comment (Kleidukos) [reply scripts/release/create-release-metadata-for-ghcup.sh]: This link talks about the Bourne Shell (`sh`) but this PR uses `bash`. On my FreeBSD machine `bash` is located elsewhere:

```
% whereis bash
bash: /usr/local/bin/bash /usr/local/man/man1/bash.1.gz
```

Comment (Kleidukos) [reply scripts/release/create-release-metadata-for-ghcup.sh]: I don't mind this script being Linux-only, but then in this case there should be a test on `uname` at the beginning of the script.

File: scripts/release/download-cabal-install-release-binaries.sh
Comment (chreekat) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: 👻

Comment (chreekat) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: You can drop the ~ and just use https://downloads.haskell.org/cabal/ as the base url. Otherwise that's right. And if you don't want maintainers to have to download things locally, once the script is in good shape we can add it to the release pipeline at gitlab.haskell.org/Haskell/cabal. I can help with that.

Comment (hasufell) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: > You can drop the ~ and just use https://downloads.haskell.org/cabal/ as the base url. 

That doesn't work. We need to download from the gitlab CI pipeline that creates the release, because we have to anyway in order to upload it to downloads.haskell.org.

File: scripts/release/create-release-metadata-for-ghcup.sh
Comment (ulysses4ever) [reply scripts/release/create-release-metadata-for-ghcup.sh]: Not every Linux has `/bin/bash` either. For example, I use NixOS, for which only the proposed change will work. I assume, Guix is the same. I highly recommend to apply the change with /usr/bin/env: from what I heard in the NixOS community, this is *much* more portable.

Comment (ulysses4ever) [reply scripts/release/create-release-metadata-for-ghcup.sh]: Oh, it was actually applied. Good.

Comment (hasufell) [reply scripts/release/create-release-metadata-for-ghcup.sh]: > from what I heard in the NixOS community, this is much more portable.

Yeah, they've also been spreading misinformation that this is recommended by posix standard :p 

As per posix: both are equally non-portable.

I believe switching to `/bin/sh` is practically more portable. Since NixOS isn't properly FHS compatible, I consider it an outlier.

Comment (arjunkathuria) [reply scripts/release/create-release-metadata-for-ghcup.sh]: if its agreed up, i can change the other script (the download script) to use `/usr/bin/env bash` as well

File: scripts/release/download-cabal-install-release-binaries.sh
Comment (chreekat) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: Oh duh. Well I think a release manager would be better qualified to answer. :) 

I'm happy to help automate later at any rate!

Comment (hasufell) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: > Oh duh. Well I think a release manager would be better qualified to answer. :)

We're asking for the URL to download artifacts from gitlab.

----

Anyway, @arjunkathuria look here: https://docs.gitlab.com/ee/api/job_artifacts.html#download-the-artifacts-archive

An example is:

```sh
curl -L 'https://gitlab.haskell.org/api/v4/projects/1164/jobs/artifacts/cabal-install-v3.10.2.1/download?job=build-aarch64-darwin' -o foo.zip
```

I'm not sure what the correct job names for the matrix builds are... this downloads a zip archive with all the individual artifacts in it that a job produces. There's also a way to just download a specific artifact from a job.

One major caveat: this only works if the pipeline as a whole succeeded.

Comment (arjunkathuria) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: This gets a bit more involved.

To get the artifacts for all platforms and architectures, from what i can read from the docs, we'd either have to give:

 
- individual job ids for each one to the script.
or
- hardcode all the job names like `build-x86_64-windows`, `build-linux: [x86_64, x86_64-linux, deb10] ` and a branch name like `main`.

unzip the `artifact.zip` archives and put the `cabal-install-*` files in a single directory.

and even then it'll work only if the pipeline succeeds for all architectures and platforms at once.

How does this happen currently, manual by-hand downloads by the release manager?

File: scripts/release/create-release-metadata-for-ghcup.sh
Comment (ulysses4ever) [reply scripts/release/create-release-metadata-for-ghcup.sh]: Posix or not, this practically covers more systems.

Comment (Bodigrim) [reply scripts/release/create-release-metadata-for-ghcup.sh]: Practically `#!/bin/sh` is the most portable option. A system might have no `bash` at all. Any particular reason not to use plain old `sh`?

File: scripts/release/download-cabal-install-release-binaries.sh
Comment (hasufell) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: Yes, it's a little more work.

Comment (arjunkathuria) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: Do you think that this is better suited as its own issue/PR pair? 
The scope of this particular PR was to make a script that generates metadata for GHCup for a cabal release.

I think we can iterate on the how-to and where-to download parts that make releases easier as its own rel-eng issue as it would involve into its own discussion/interface requirements.

Comment (chreekat) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: > We're asking for the URL to download artifacts from gitlab.

That's why I suggest talking to a past/current Cabal release manager. I imagine they already have a workflow for this. While I can point to release pipelines and explain in general how to find the artifacts, it would be better to know what the actual current process is. Maybe @Kleidukos can help.

> Do you think that this is better suited as its own issue/PR pair?
The scope of this particular PR was to make a script that generates metadata for GHCup for a cabal release.

Yeah, I totally agree with that.

To actually automate it, it would be a lot easier to run the script in the GitLab pipeline that generates the bindists in the first place. That's what I can help with. https://github.com/haskell/cabal/blob/master/.gitlab-ci.yml is the entrypoint.

Comment (hasufell) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: I think the best course of action is to purge gitlab CI.

Comment (Kleidukos) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: You may want to re-purpose the tools of the GHC release engineering team: 

* https://gitlab.haskell.org/ghc/ghc/-/tree/master/.gitlab/rel_eng/mk-ghcup-metadata?ref_type=heads

* https://gitlab.haskell.org/ghc/ghc/-/tree/master/.gitlab/rel_eng/fetch-gitlab-artifacts?ref_type=heads

Comment (arjunkathuria) [reply scripts/release/download-cabal-install-release-binaries.sh:31]: Yea, they do it in two steps.

1. get the job ids from a pipeline id.
2. try to get artifacts from these job ids and error if anyone of those failed.
3. If successful, generate metadata from these artifacts.

i suppose i try repurposing the logic in the shell script here. 
If that's not required or if that needs to be a separate issue/PR pair, please do let me know.

**Comment by hasufell** [comment]
I migrated release CI to github and also added the yaml snippet and download scripts. That makes this PR obsolete.

https://github.com/haskell/cabal/pull/9437

**Comment by Mikolaj** [comment]
You mean, "obsolete in the future"? That's what all PRs are. :)

I'm guessing (@Kleidukos, correct me if I'm wrong), in the best case, we'd migrate to exclusively github for release after a couple of release using both gitlab and github and fine-tuning the new (github) method. For these couple of releases, this PR is still likely to be a great help (and possibly for the fine-tuning).

Still, it's fair heads-up to @arjunkathuria, in case this PR needs a lot of work still and, presumably, it's life-span would be shorter than initially expected. So, @arjunkathuria, it's your call and, in any case, the work you spent on that is very much appreciated.

**Comment by arjunkathuria** [comment]
Thanks @Mikolaj 

Ah well.
This PR is ready if we download from `download.haskell.org` and just want to generate the GHCup metadata, which i believe was the original scope.

I'd be alright to close this when https://github.com/haskell/cabal/pull/9437 gets merged, as it probably supersede this.

I'd like to thank everyone involved  @Mikolaj @hasufell @Kleidukos @chreekat @ulysses4ever and @Bodigrim for their time and effort put into providing great feedback and reviewing this.

**Comment by hasufell** [comment]
> just looked at #9437 to find that @hasufell closed it? unsure where this leaves this PR, any inputs @Mikolaj @Kleidukos @hasufell ?
> 
> i'd be pretty alright to close this if the script isn't required/useful anymore or if the underlying assumptions and scope changed.

I won't be using the bindists from cabal CI for ghcup, except for the `ghcup-vanilla-0.0.8.yaml` channel, which is still up to upstream to contribute. So your script may still be useful for cabal devs.

**Comment by Mikolaj** [comment]
@arjunkathuria: It seems we are going to have many releases soon, so your script is likely to be very useful, but the current release manager (@Kleidukos) would know best.

**Comment by arjunkathuria** [comment]
Bump.

cc: @Kleidukos

**Comment by Kleidukos** [comment]
Ok let's bring this in.
@arjunkathuria Thanks again for this work, this is going to be most valuable.

**Review by Kleidukos — APPROVED** [review]

**Review by Mikolaj — APPROVED** [review]
This will, obviously, be refined on actual usage, but that's a solid start at least. Thank you!

**Comment by Mikolaj** [comment]
[BTW, I haven't informed the git police yet, but the conversion is that commit messages are sentences in imperative present tense without the ending dot ("Tweak wibbles in fobuscator").]

**Comment by Mikolaj** [comment]
@arjunkathuria: could you mark this PR as "Maintainers are allowed to edit this pull request. " so that mergify can rebase and merge it? Let me know if that's problematic.

**Comment by arjunkathuria** [comment]
Hi @Mikolaj, hope you've been well.

>Let me know if that's problematic.

Hardly. I just rebased to latest to save you some time : D

**Comment by Mikolaj** [comment]
Great. That was enough for mergify. Thank you again for your contribution!

**Pull request merged**
